File: app/src/main/java/org/eu/droid_ng/wellbeing/prefs/DayPicker.java
package org.eu.droid_ng.wellbeing.prefs;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.ToggleButton;

import androidx.annotation.Nullable;

import org.eu.droid_ng.wellbeing.R;

import java.util.Calendar;
import java.util.function.Consumer;

public class DayPicker extends FrameLayout {
	private ToggleButton[] views;
	private boolean[] values = new boolean[7]; // Monday -> Sunday like Java DayOfWeek
	private int firstDayOfWeek;
	private Consumer<boolean[]> onValuesChangeListener;

	public DayPicker(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
		initView();
	}

	public DayPicker(Context context, AttributeSet attrs) {
		super(context, attrs);
		initView();
	}

	public DayPicker(Context context) {
		super(context);
		initView();
	}

	private void initView() {
		inflate(getContext(), R.layout.dpicker, this);
		ToggleButton day1 = findViewById(R.id.dayPickerDay1);
		ToggleButton day2 = findViewById(R.id.dayPickerDay2);
		ToggleButton day3 = findViewById(R.id.dayPickerDay3);
		ToggleButton day4 = findViewById(R.id.dayPickerDay4);
		ToggleButton day5 = findViewById(R.id.dayPickerDay5);
		ToggleButton day6 = findViewById(R.id.dayPickerDay6);
		ToggleButton day7 = findViewById(R.id.dayPickerDay7);
		views = new ToggleButton[] {day1, day2, day3, day4, day5, day6, day7};

		firstDayOfWeek = (Calendar.getInstance().getFirstDayOfWeek() - 2) % 7;
		for (int i = 0; i < 7; i++) {
			ToggleButton v = views[i];
			int textToSet;
			int javaDayOfWeek = Math.floorMod((i + firstDayOfWeek), 7);
			switch (javaDayOfWeek + 2) {
				case Calendar.MONDAY:
					textToSet = R.string.dpicker_monday;
					break;
				case Calendar.TUESDAY:
					textToSet = R.string.dpicker_tuesday;
					break;
				case Calendar.WEDNESDAY:
					textToSet = R.string.dpicker_wednesday;
					break;
				case Calendar.THURSDAY:
					textToSet = R.string.dpicker_thursday;
					break;
				case Calendar.FRIDAY:
					textToSet = R.string.dpicker_friday;
					break;
				case Calendar.SATURDAY:
					textToSet = R.string.dpicker_saturday;
					break;
				case Calendar.SUNDAY:
				default:
					textToSet = R.string.dpicker_sunday;
					break;
			}
			CharSequence textToSet2 = getContext().getString(textToSet);
			v.setTextOn(textToSet2);
			v.setTextOff(textToSet2);
			v.setOnCheckedChangeListener((view, isChecked) -> {
				values[javaDayOfWeek] = isChecked;
				if (onValuesChangeListener != null) {
					onValuesChangeListener.accept(values);
				}
			});
		}

		setValues(values);
	}

	public void setValues(boolean[] values) {
		this.values = values;
		for (int i = 0; i < 7; i++) {
			ToggleButton v = views[i];
			int javaDayOfWeek = Math.floorMod((i + firstDayOfWeek), 7);
			v.setChecked(values[javaDayOfWeek]);
		}
	}

	public boolean[] getValues() {
		return values;
	}

	public void setOnValuesChangeListener(@Nullable Consumer<boolean[]> onValuesChangeListener) {
		this.onValuesChangeListener = onValuesChangeListener;
	}
}

File: app/src/main/java/org/eu/droid_ng/wellbeing/prefs/DayPicker.kt
package org.eu.droid_ng.wellbeing.prefs

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatToggleButton
import org.eu.droid_ng.wellbeing.R
import java.util.Calendar
import java.util.function.Consumer

class DayPicker(context: Context, attrs: AttributeSet?, defStyle: Int)
	: FrameLayout(context, attrs, defStyle) {
	private var views: Array<AppCompatToggleButton>
	var values = BooleanArray(7) // Monday -> Sunday like Java DayOfWeek
		set(value) {
			field = value
			for (i in 0..6) {
				val v = views[i]
				val javaDayOfWeek = Math.floorMod((i + firstDayOfWeek), 7)
				v.isChecked = field[javaDayOfWeek]
			}
		}
	private var firstDayOfWeek = 0
	private var onValuesChangeListener: Consumer<BooleanArray>? = null

	constructor(context: Context, attrs: AttributeSet?) : this(
		context, attrs, 0
	)
	constructor(context: Context) : this(context, null)

	init {
		inflate(context, R.layout.dpicker, this)
		val day1 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay1)
		val day2 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay2)
		val day3 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay3)
		val day4 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay4)
		val day5 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay5)
		val day6 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay6)
		val day7 = findViewById<AppCompatToggleButton>(R.id.dayPickerDay7)
		views = arrayOf(day1, day2, day3, day4, day5, day6, day7)

		firstDayOfWeek = (Calendar.getInstance().firstDayOfWeek - 2) % 7
		for (i in 0..6) {
			val v = views[i]
			var textToSet: Int
			val javaDayOfWeek = Math.floorMod((i + firstDayOfWeek), 7)
			textToSet = when (javaDayOfWeek + 2) {
				Calendar.MONDAY -> R.string.dpicker_monday
				Calendar.TUESDAY -> R.string.dpicker_tuesday
				Calendar.WEDNESDAY -> R.string.dpicker_wednesday
				Calendar.THURSDAY -> R.string.dpicker_thursday
				Calendar.FRIDAY -> R.string.dpicker_friday
				Calendar.SATURDAY -> R.string.dpicker_saturday
				Calendar.SUNDAY -> R.string.dpicker_sunday
				else -> R.string.dpicker_sunday
			}
			val textToSet2: CharSequence = context.getString(textToSet)
			v.textOn = textToSet2
			v.textOff = textToSet2
			v.setOnCheckedChangeListener { _, isChecked ->
				values[javaDayOfWeek] = isChecked
				if (onValuesChangeListener != null) {
					onValuesChangeListener!!.accept(values)
				}
			}
		}

		this.values = values // call setter
	}
	fun setOnValuesChangeListener(onValuesChangeListener: Consumer<BooleanArray>?) {
		this.onValuesChangeListener = onValuesChangeListener
	}
}